### PR TITLE
BUG: Fix weak promotion with some mixed float/int dtypes

### PR DESCRIPTION
**Pull request by seberg**
This reorganizes/simplifies promotion a bit, but unfortunately this means moving the promotion of our concrete DTypes with the abstract "weak" ones into the concrete DTypes making the order of "who deals with whom" fully defined by:

    complex' > floats > pycomplex > pyfloat > ints > pyint > bool

where py<...> is the DType corresponding to the Python type. There are two points here:
* "inexact" is the meaningful category not complex
* previously, the abstract DTypes `pyfloat`, etc. knew about the concrete ones and there was some special handling for that. (as mentioned above, this is gone now)

Why was it this complicated?!  I think the initial code wanted to deal with value-based promotion and the abstract DTypes could have had a value attached to them.
There was thus more code later on, dealing with abstract DTypes more concretely (rather than identical to the rest).

This is now gone, concrete DTypes promote basically the same as their concrete counter-parts.  This makes the promotion logic simpler (and I think still just as correct), although the concrete type promotion now needs to know about it.

### CODE REVIEW

**Comment by tylerjereddy** [comment]
A bit over my head, but I did check that this branch against SciPy `main` doesn't change our full test suite result with `NPY_PROMOTION_STATE=weak` from the one reported at https://github.com/scipy/scipy/issues/19239

**Comment by lezcano** [comment]
@ev-br can you have a look at this one and see if we need to change anything in our implementation?

**Comment by seberg** [comment]
Can we please review/merge this?  It's a clear bug fix in something that is vital to move forward.  (not that it is a _big_ thing in that, but still)

File: numpy/core/src/multiarray/common_dtype.c [hunk 1]
```diff
@@ -90,26 +90,17 @@ PyArray_CommonDType(PyArray_DTypeMeta *dtype1, PyArray_DTypeMeta *dtype2)
...
- *
- * The reason why this may be acceptable for abstract DTypes, is that
- * the value-dependent abstract DTypes may provide default fall-backs.
- * The priority inversion effectively means that abstract DTypes are ordered
- * just below their concrete counterparts.
- * (This fall-back is convenient but not perfect, it can lead to
- * non-minimal promotions: e.g. `np.uint24 + 2**20 -> int32`. And such
- * cases may also be possible in some mixed type scenarios; they can be
- * avoided by defining the promotion explicitly in the user DType.)
+ * For the moment, abstract dtypes are not handled special here.  In a first
```
Comment (mattip) [inline numpy/core/src/multiarray/common_dtype.c:98]:
```suggestion
 * For the moment, abstract dtypes are not handled special here.  In a first
```
```suggestion
 * Abstract dtypes are not handled specially here.  In a first
```

File: numpy/core/src/multiarray/common_dtype.c [hunk 2]
```diff
@@ -90,26 +90,17 @@ PyArray_CommonDType(PyArray_DTypeMeta *dtype1, PyArray_DTypeMeta *dtype2)
...
+ * version they were but this version also tried to be able to do value-based
+ * behavior.
+ * There may be some advantage to special casing the abstract ones (e.g.
+ * so that the concrete ones do not have to deal with it), but this would
+ * require more complex handling later on.
```
Comment (mattip) [inline numpy/core/src/multiarray/common_dtype.c:103]:
```suggestion
 * require more complex handling later on. See the logic in
 * default_builtin_common_dtype
```

**Review by mattip — APPROVED** [review]
LGTM, a few small documentation nits. I will merge soon if you want to ignore them.

**Comment by mattip** [comment]
Thanks @seberg

**Pull request merged**